docs/source/enterprise-guide/graph-security.mdx

```diff
@@ -0,0 +1,262 @@
+
+```graphql
+query {
+  authors(first: 100) {
+    name
+    posts(last: 1000) {
+      title
+      content
+    }
+  }
+}
+```
+
+When paginating fields, it's important to set a maximum number of items that may be returned in a single response. In the example above, you may wish to return a GraphQL error when executing the `posts` field resolver instead of attempting to return 1000 posts for each of the 100 authors.
+
+And as mentioned in the [federated schema design section](/enterprise-guide/federated-schema-design/), with federation, it’s a best practice to consider how you standardize these pagination characteristics of your fields across subgraphs so that client developers have a predictable experience querying different fields of the API. That’s not to say that all fields should be paginated in the same way (for example, only using cursor-based pagination or only using offset-based pagination), but their arguments and their outputs should provide an intuitive experience on the client-side.
+
+#### Validate and sanitize data
+
+Validating and sanitizing client-submitted data is important for any API, and a federated data graph is no exception. In general, the usual rules for validation and sanitization of untrusted inputs will apply to GraphQL when resolving fields based on user-provided inputs. And as previously discussed, when users supply invalid values as operation arguments the resulting errors should provide as little details as possible in production environments.
+
+A well-designed GraphQL schema can also help guard against injection attacks by codifying validation and sanitization directly into types. For example, enum values can limit the range of what may be submitted for argument values and custom scalars or directives may be used to validate, escape, or normalize values as well. However, custom scalars should be handled with care as their misuse may create other vulnerabilities, such as [a `JSON` scalar type enabling a NoSQL injection attack](http://www.petecorey.com/blog/2017/06/12/graphql-nosql-injection-through-json-types/).
+
+#### Set timeouts
+
+Timeouts are another useful tool for stopping GraphQL operations that are consuming more server resources than expected. With federation, timeouts may commonly be applied at three different levels. At the highest level, a timeout may be set on the gateway's HTTP server (or an idle timeout on a load balancer in front of it).
+
+The second level where you may configure timeouts is on the gateway's requests to subgraph services. This can be done by defining a custom fetcher for `ApolloGateway` in a `RemoteGraphQLDataSource` as follows:
+
+```js
+const gateway = new ApolloGateway({
+  // ...
+  buildService({ name, url }) {
+    const fetcher = (input, init) => {
+      if (init) {
+        init.timeout = 3000;
+      } else {
+        init = { timeout: 3000 };
+      }
+      return fetch(input, init);
```

Comment:
Is it left to an exercise to the user to provide this from some Fetch API?  (I don't see it noted that it's an exercise to the user? 🤔 )

Reply:
Thanks for flagging. I've added import lines to improve clarity.